### PR TITLE
Promotion edit: Add "Applies To" for products

### PR DESCRIPTION
**Pull request by coderkevin**
This adds the UI code necessary to implement a filtered "Applies To" selection for a single product for a product sale.

To Test:
1. `npm start`
2. Visit `http://calypso.localhost:3000/store/promotion/<site url>`
3. Select "Individual Product Sale"
4. Test action in "Applies to product" field at bottom (search filtering, selection, etc.)
5. Select one of the coupon options.
6. Try out Applies to "all"
8. Try out Applies to "Specific Products"

applies-to-product

### CODE REVIEW

**Comment by kellychoffman** [comment]
Awesome. Just some styling fixes but I'll leave that to Jay.

**Comment by coderkevin** [comment]
@jameskoster Thanks for the css changes. I did one correction though. I moved the padding to the label because it was reducing the clickable area for each row when it was in the div.

**Comment by allendav** [comment]
Suggestion - sort the products by name, alphabetically to make individual products easier to find.

cc @kellychoffman @jameskoster

**Comment by allendav** [comment]
I noticed if I enter a space (e.g. after the word product in the example below) then the blue outline appears. Would be nice to avoid that

screen shot 2017-10-27 at 9 27 34 am

**Comment by kellychoffman** [comment]
Styling looks good.

> Suggestion - sort the products by name, alphabetically, to make individual products easier to find.

Agree!

File: client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import FormRadio from 'components/forms/form-radio';
+import { getProductCategories } from 'woocommerce/state/sites/product-categories/selectors';
+import { getPromotionableProducts } from 'woocommerce/state/selectors/promotions';
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import Search from 'components/search';
+
+function productContainsString( product, textString ) {
+	const matchString = textString.trim().toLowerCase();
+
+	if ( -1 < product.name.toLocaleLowerCase().indexOf( matchString ) ) {
```
Comment (allendav) [inline client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: Just curious why toLocalLowerCase wasn't used to generate matchString as well?

Comment (allendav) [reply client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: Could just return the conditional

File: client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		const { searchFilter } = this.state;
+
+		if ( 0 === searchFilter.length ) {
+			return products;
+		}
+
+		return products && products.filter(
+			( product ) => {
+				return productContainsString( product, searchFilter ) ||
+					isProductSelected( value, product.id );
```
Comment (allendav) [inline client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: I like how we're keeping the selected product - the UX is less confusing that way

**Comment by coderkevin** [comment]
> I noticed if I enter a space (e.g. after the word product in the example below) then the blue outline appears. Would be nice to avoid that

Nice find! This appears to be a problem with the Calypso search component and it happens everywhere. I've logged an issue for that here: https://github.com/Automattic/wp-calypso/issues/19283

**Comment by coderkevin** [comment]
> Suggestion - sort the products by name, alphabetically, to make individual products easier to find.

Problem is, in our product list we go with the default search. I feel if we change to sort by alpha here, we should probably sort by alpha there too?

File: client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js (hunk 1 above)
Comment (coderkevin) [reply client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: I noticed that too, but I think I changed it in the next PR to use `toLocaleLowerCase()` on both.

Comment (coderkevin) [reply client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: > Could just return the conditional
Yeah, I suppose... I guess I get nervous about returning numerics as booleans.

File: client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js (hunk 2 above)
Comment (coderkevin) [reply client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js]: Yeah, it seemed the least confusing way to do it, but I think it's still just a little confusing.

**Comment by coderkevin** [comment]
Adjusted the margins of the search box and type selector.

**Comment by coderkevin** [comment]
Rebased onto current master.

**Comment by coderkevin** [comment]
Merging #19253 and #19255 into this PR before merging to master.

**Pull request merged**